src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -0,0 +1,1599 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The exported function names need to be unique (can't be disambiguated based on signature), hence
+// we introduce suffix letters to indicate the general patterns used.
+// * A suffix means aligned and padded for SSE operations.
+// * U suffix means unaligned and unpadded.
+// * P suffix means sparse (unaligned) partial vector - the vector is only part of a larger sparse vector.
+// * Tran means the matrix is transposed.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class AvxIntrinsics
+    {
+        private const int Vector256Alignment = 32;
+
+        private static bool Compat(AlignedArray a)
+        {
+            Contracts.AssertValue(a);
+            Contracts.Assert(a.Size > 0);
+            return a.CbAlign == Vector256Alignment;
+        }
+
+        private static unsafe float* Ptr(AlignedArray a, float* p)
```

Comment:
`GetPointer`? And use a more descriptive name than `a` and `p`?

Reply:
Used: `GetBaseAddress`, `alignedArray`, and `address`.  The `q` that is returned is now known as `baseAddress`.

Reply:
I think the name of `address` is still confusing. Looking at the code, it is being used as a `base` (`a.GetBase()` is added to it).

We have a lot of concepts of "base" here, with not much clarification. We have an array (which we call `GetBase`) for, a pointer (which is used as the base address that `GetBase` is added onto), and we are returning a new base address.

Reply:
Suggestion: `p -> startAddress`, `q -> baseAddress`?

Reply:
or `GetAlignedBase`, `p -> unalignedBase`, `q -> alignedBase`?